### PR TITLE
add functions for generating prompt_str and prepare inputs for training

### PR DESCRIPTION
**Pull request by khai-meetkai**

### CODE REVIEW

File: functionary/prompt_utils.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        },
+        {"role": "function", "content": "{'price': {'price': '$20000'}}", "name": "get_car_price"},
+        {
+            "role": "assistant",
+            "content": "No, the car Tang is less expensive than the car Song. The car Song is priced at $25,000, while the car Tang is priced at $20,000.",
+        },
+    ]
+
+
+def test_prompt():
```
Comment (musab-mk) [inline functionary/prompt_utils.py:165]: Lets put this into tests

File: functionary/prompt_utils.py
Comment (musab-mk) [inline functionary/prompt_utils.py:1]: Lets add docstrings to the functions

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: tests/test_prompt_creation.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        },
+    ]
+
+
+def test_prompt():
+    tokenizer = LlamaTokenizer.from_pretrained("musabgultekin/functionary-7b-v1")
+    # first we add stop_tokens to the tokenizer
+    added_tokens = [e.value for e in EndToken]
+    print("added_tokens: ", added_tokens)
+    tokenizer.add_tokens(added_tokens)
```
Comment (musab-mk) [inline tests/test_prompt_creation.py:36]: Since these are special tokens, we should use add_special_tokens. Like Jeffrey used

File: functionary/prompt_utils.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    elif message["role"] == "user":
+        text = "user:\n{content}\n".format(content=content)
+
+    elif message["role"] == "assistant":
+        function = None
+        arguments = None
+        if (
+            "function_call" in message
+        ):  # format of openai: {"role": assistant, "function_call": {"name": xxx, "arguments": xxx}}
+            assert content is None
```
Comment (musab-mk) [inline functionary/prompt_utils.py:72]: We can have content too on function calls. For example:
"Sure! Lets get the weather for you!"

Comment (musab-mk) [reply functionary/prompt_utils.py:72]: This exists on gpt3.5 function calls too

File: functionary/prompt_utils.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            function = message["function_call"]["name"]
+            arguments = message["function_call"]["arguments"] + stop_token
+            text = f"assistant to={function}:\n{arguments}\n"
+        elif content is not None:  # this is text content
+            text = f"assistant:\n{content}\n"
+        else:  # if no function call and content is None --> this is used at inference
+            text = "assistant"
+
+    # This chunk of code below is to assert that stop_token is added correctly
```
Comment (musab-mk) [inline functionary/prompt_utils.py:81]: This is basically tests. So instead of assertions, lets use seperate tests to make sure its working as we expected. We need a test for each role and for each combination.

Comment (musab-mk) [reply functionary/prompt_utils.py:81]: This is really important. Cause there are lots of edge-cases as you probably know

File: functionary/prompt_utils.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    # where current_stop_token is EndToken.assistant or EndToken.function_call
+    for index, tok_id in enumerate(input_token_ids):
+        if tok_id in id_to_endtoken:  # Find position of end_token in final_prompt
+            end_token = id_to_endtoken[tok_id]
+            if (
+                end_token == EndToken.assistant or end_token == EndToken.function_call
+            ):  # only compute loss from tokens of assistant
+                for i in range(start + 2, index + 1):  # The reason for start + 2 is to ignore: "\nassistant" (2 tokens)
+                    labels[i] = input_token_ids[i]  # overwrite -100 to compute the loss
+                if verbose:
```
Comment (musab-mk) [inline functionary/prompt_utils.py:197]: We need to have tests for this instead of manually checking. For the whole prepare_training_inputs

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: functionary/prompt_utils.py (hunk 3 above)
Comment (khai-meetkai) [reply functionary/prompt_utils.py:72]: Oh really, because from my test haven't seen this

File: functionary/prompt_utils.py (hunk 4 above)
Comment (khai-meetkai) [reply functionary/prompt_utils.py:81]: Sure will check this in test

File: tests/test_prompt_creation.py (hunk 2 above)
Comment (khai-meetkai) [reply tests/test_prompt_creation.py:168]: Done

File: functionary/prompt_utils.py (hunk 5 above)
Comment (khai-meetkai) [reply functionary/prompt_utils.py:181]: I have implemented unittest at: tests/test_prompt_creation.py

File: functionary/prompt_utils.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+class EndToken(str, Enum):
+    system = "<|END_OF_SYSTEM|>"
+    user = "<|END_OF_USER|>"
+    assistant = "<|END_OF_ASSISTANT|>"
+    function = "<|END_OF_FUNCTION_RESULT|>"
+    function_call = "<|END_OF_FUNCTION_CALL|>"
+
+
+def get_end_token_for_message(message: Dict) -> EndToken:
```
Comment (musab-mk) [inline functionary/prompt_utils.py:14]: I believe we can make this @staticmethod of EndToken. 
Something like: from_message

File: functionary/prompt_utils.py
Comment (musab-mk) [inline functionary/prompt_utils.py:1]: Can we rename this file to prompt.py ?
"_utils" doesnt add that much value

File: tests/test_prompt_creation.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            )  # note that need to add: "\nassistant" because we mask this, see line 194 in prompt_utils.py
+            prompt = get_text_from_message(message)
+            # decoded_content and prompt should be the same
+            # to avoid any mistakes of tokenizer like adding white space we will compare after removing space
+            self.assertEqual(
+                re.sub("\s", "", decoded_content),
+                re.sub("\s", "", prompt),
+                "decoded content is different from original content",
+            )
+
```
Comment (musab-mk) [inline tests/test_prompt_creation.py:137]: can we add a test that checks exact equality of an input and output of "get_prompt_from_messages"

**Review by musab-mk — CHANGES_REQUESTED** [review]

File: tests/test_prompt_creation.py (hunk 7 above)
Comment (khai-meetkai) [reply tests/test_prompt_creation.py:137]: Sure

File: functionary/prompt_utils.py
Comment (khai-meetkai) [reply functionary/prompt_utils.py:1]: Sure !

File: functionary/prompt_utils.py (hunk 6 above)
Comment (khai-meetkai) [reply functionary/prompt_utils.py:14]: will change to static method of EndToken now

**Review by musab-mk — APPROVED** [review]
